kubernetes.go

```diff
@@ -257,28 +248,75 @@ func (c *KubernetesClient) Watch() error {
 	})
 	go updater.Run()
 
+	c.nodeStore = NodeStore{NewLocalStore()}
+	c.serviceStore = ServiceStore{NewLocalStore()}
+	c.endpointsStore = EndpointsStore{NewLocalStore()}
+
+	updateStore := func(s Store, e watch.Event, equal EqualFunc) {
+		if e.Object == nil {
+			return
+		}
+		switch e.Type {
+		case watch.Added, watch.Modified:
+			// We handle Added the same as Modified because when reconnecting we
+			// receive everything as Added
+			old := s.Update(e.Object)
+			if old != nil {
+				eq, err := equal(old, e.Object)
+				if err != nil {
+					log.Println(err)
+					return
+				}
+				if eq {
+					return
+				}
+			}
+			updater.Signal()
+		case watch.Deleted:
+			s.Delete(e.Object)
+			updater.Signal()
+		}
+	}
+
 	var more bool
 	var e watch.Event
 	for {
 		select {
 		case e, more = <-c.nodeWatcher.ResultChan():
-			if e.Type == watch.Added || e.Type == watch.Deleted {
-				updater.Signal()
-			}
-		case _, more = <-c.serviceWatcher.ResultChan():
-			updater.Signal()
+			updateStore(c.nodeStore, e, EqualUIDs)
+		case e, more = <-c.serviceWatcher.ResultChan():
+			updateStore(c.serviceStore, e, func(a, b runtime.Object) (bool, error) {
+				eqUIDs, err := EqualUIDs(a, b)
+				if err != nil {
+					return false, err
+				}
+				eqVersion, err := EqualResourceVersion(a, b)
+				if err != nil {
+					return false, err
+				}
+				return eqUIDs && eqVersion, nil
+			})
 		case e, more = <-c.endpointsWatcher.ResultChan():
```

Comment:
Changes in any endpoint will trigger a reload, even if they below to a not exposed service. This is not really a regression, it is also happening in current version, to be fixed in other PR.